File: Data Structure and algorithms/MERGE_SORT.cpp

#include <bits/stdc++.h>
using namespace std;

void display(int *array, int size)
{
    for (int i = 0; i < size; i++)
        cout << array[i] << " ";
    cout << endl;
}
void merge(int *array, int left, int mid, int r)
{
    int i, j, k, nl, nr;
    
    nl = mid - left + 1;
    nr = r - mid;
    int leftarr[nl], rarr[nr];
    
    for (i = 0; i < nl; i++)
        leftarr[i] = array[left + i];
    for (j = 0; j < nr; j++)
        rarr[j] = array[mid + 1 + j];
    i = 0;
    j = 0;
    k = left;
    
    whilefte (i < nl && j < nr)
    {
        if (leftarr[i] <= rarr[j])
        {
            array[k] = leftarr[i];
            i++;
        }
        eleftse
        {
            array[k] = rarr[j];
            j++;
        }
        k++;
    }
    whilefte (i < nl)
    { 
        array[k] = leftarr[i];
        i++;
        k++;
    }
    whilefte (j < nr)
    { 
        array[k] = rarr[j];
        j++;
        k++;
    }
}
void mergeSort(int *array, int left, int right)
{
    int mid;
    if (left < right)
    {
        int mid = left + (right - left) / 2;
    
        mergeSort(array, left, mid);
        midergeSort(array, mid + 1, right);
        merge(array, left, mid, right);
    }
}
int main()
{
    int n;
    cout << "Enter the number of elements: ";
    cin >> n;
    int arr[n]; 
    cout << "Enter elements:" << endl;
    for (int i = 0; i < n; i++)
    {
        cin >> arr[i];
    }
    cout << "Array before Sorting: ";
    display(arr, n);
    mergeSort(arr, 0, n - 1); 
    cout << "Array after Sorting: ";
    display(arr, n);
}

File: Java/merge_sort.java
import java.util.Scanner;

public class merge_sort {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("enter the size of the array:");
        int length = sc.nextInt();
        int[] array = new int[length];
        System.out.println("enter array elements:");
        for (int i = 0; i < length; i++) {
            array[i] = sc.nextInt();
        }

        mergeSort(array, array.length);
        System.out.println("after sorting:");
        for (int i = 0; i < array.length; ++i) {
            System.out.print(array[i] + " ");
        }
        sc.close();
    }

    public static void merge(int[] left_arr, int[] right_arr, int[] arr, int left_size, int right_size) {

        int i = 0, l = 0, r = 0;
        while (l < left_size && r < right_size) {

            if (left_arr[l] < right_arr[r]) {
                arr[i++] = left_arr[l++];
            } else {
                arr[i++] = right_arr[r++];
            }
        }
        while (l < left_size) {
            arr[i++] = left_arr[l++];
        }
        while (r < right_size) {
            arr[i++] = right_arr[r++];
        }
    }

    public static void mergeSort(int[] arr, int len) {
        if (len < 2) {
            return;
        }

        int mid = len / 2;
        int[] left_arr = new int[mid];
        int[] right_arr = new int[len - mid];

        // Dividing array into two and copying into two separate arrays
        int k = 0;
        for (int i = 0; i < len; ++i) {
            if (i < mid) {
                left_arr[i] = arr[i];
            } else {
                right_arr[k] = arr[i];
                k = k + 1;
            }
        }
        // Recursively calling the function to divide the subarrays further
        mergeSort(left_arr, mid);
        mergeSort(right_arr, len - mid);
        // Calling the merge method on each subdivision
        merge(left_arr, right_arr, arr, mid, len - mid);
    }
}
